lib/icingadb/icingadb-worker.cpp

```diff
@@ -0,0 +1,503 @@
+		}
+	}
+}
+
+/**
+ * Dequeue and process a single pending item.
+ *
+ * This function processes a single pending item from the pending items container. It iterates over
+ * the items in insertion order and checks if the first item is old enough to be processed (at least
+ * 1000ms old) unless we're being shutting down. If the item can be processed, it attempts to acquire
+ * a lock on the associated config object (if applicable) and processes the item accordingly.
+ *
+ * If the item cannot be processed right now because it's too new, the function returns a duration
+ * indicating how long to wait before retrying. Also, if no progress was made during this iteration
+ * (i.e., no item was processed), it returns a short delay to avoid busy-looping.
+ *
+ * @param lock A unique lock on the pending items mutex (must be acquired before calling this function).
+ *
+ * @return A duration indicating how long to wait before retrying.
+ */
+std::chrono::duration<double> IcingaDB::DequeueAndProcessOne(std::unique_lock<std::mutex>& lock)
+{
+	using namespace std::chrono_literals;
+	namespace ch = std::chrono;
+
+	bool madeProgress = false; // Did we make any progress in this iteration?
+	ch::duration<double> retryAfter{0}; // If we can't process anything right now, how long to wait before retrying?
+	auto now = ch::steady_clock::now();
+
+	auto& seqView = m_PendingItems.get<1>();
+	for (auto it(seqView.begin()); it != seqView.end(); ++it) {
+		if (it != seqView.begin()) {
+			if (std::holds_alternative<RelationsDeletionItem>(*it)) {
+				// We don't know whether the previous items are related to this deletion item or not,
+				// thus we can't just process this right now when there are older items in the queue.
+				// Otherwise, we might delete something that is going to be updated/created.
+				break;
+			}
+		}
+
+		auto age = now - std::visit([](const auto& item) { return item.EnqueueTime; }, *it);
+		if (GetActive() && 1000ms > age) {
+			if (it == seqView.begin()) {
+				retryAfter = 1000ms - age;
+			}
+			break;
+		}
+
+		ConfigObject::Ptr cobj;
+		if (auto* citem = std::get_if<PendingConfigItem>(&*it); citem) {
+			cobj = citem->Object;
+		}
+
+		ObjectLock olock(cobj, std::defer_lock);
+		if (cobj && !olock.TryLock()) {
+			continue; // Can't lock the object right now, try the next one.
+		}
+
+		PendingItemVariant itemToProcess = *it;
+		seqView.erase(it);
+		madeProgress = true;
+
+		lock.unlock();
+		try {
+			std::visit([this](const auto& item) { ProcessPendingItem(item); }, std::move(itemToProcess));
+		} catch (const std::exception& ex) {
+			Log(LogCritical, "IcingaDB")
+				<< "Exception while processing pending item of type index '" << itemToProcess.index() << "': "
+				<< DiagnosticInformation(ex, GetActive());
+		}
+		lock.lock();
+		break;
+	}
+
+	if (!madeProgress && retryAfter == 0ms) {
+		// We haven't made any progress, so give it a short delay before retrying.
+		retryAfter = 10ms;
+	}
+	return retryAfter;
```

Comment:
I believe that would be easier to understand if you just returned an appropriate value where you currently `break;`. For example, instead of setting `madeProgress` and the last `break;`, you could just `return 0;` (± correct type) there.